plugins/inputs/cloud_pubsub_push/pubsub_push_test.go

```diff
@@ -0,0 +1,244 @@
+		},
+		{
+			name:    "post invalid data format",
+			method:  "POST",
+			path:    "/",
+			maxsize: 500 * 1024 * 1024,
+			status:  http.StatusBadRequest,
+			body:    strings.NewReader(`{"message":{"attributes":{"deviceId":"myPi","deviceNumId":"2808946627307959","deviceRegistryId":"my-registry","deviceRegistryLocation":"us-central1","projectId":"conference-demos","subFolder":""},"data":"bm90IHZhbGlkIGZvcm1hdHRlZCBkYXRh","messageId":"204004313210337","message_id":"204004313210337","publishTime":"2018-09-14T19:22:54.587Z","publish_time":"2018-09-14T19:22:54.587Z"},"subscription":"projects/conference-demos/subscriptions/my-subscription"}`),
+		},
+		{
+			name:    "post invalid structured body",
+			method:  "POST",
+			path:    "/",
+			maxsize: 500 * 1024 * 1024,
+			status:  http.StatusBadRequest,
+			body:    strings.NewReader(`{"message":{"attributes":{"thing":1},"data":"bm90IHZhbGlkIGZvcm1hdHRlZCBkYXRh"},"subscription":"projects/conference-demos/subscriptions/my-subscription"}`),
+		},
+	}
+
+	wg := &sync.WaitGroup{}
+	for _, test := range tests {
+		req, err := http.NewRequest(test.method, test.path, test.body)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		rr := httptest.NewRecorder()
+		pubPush := PubSubPush{
+			Path: "/",
+			MaxBodySize: internal.Size{
+				Size: test.maxsize,
+			},
+			sem: make(chan struct{}, 1),
+		}
+		pubPush.ctx, pubPush.cancel = context.WithCancel(context.Background())
+		pubPush.ReadTimeout = internal.Duration{
+			Duration: time.Second * 3,
+		}
+
+		if test.fail {
+			// fill buffer with fake message
+			pubPush.sem <- struct{}{}
+			pubPush.cancel()
+		}
+		if test.timeout {
+			pubPush.ReadTimeout = internal.Duration{
+				Duration: time.Second * 0,
+			}
+		}
+
+		p, _ := parsers.NewParser(&parsers.Config{
+			MetricName: "cloud_pubsub_push",
+			DataFormat: "influx",
+		})
+		pubPush.SetParser(p)
+
+		dst := make(chan telegraf.Metric, 1)
+		ro := models.NewRunningOutput("test", &testOutput{failWrite: test.fail}, &models.OutputConfig{}, 1, 1)
+		pubPush.acc = agent.NewAccumulator(&testMetricMaker{}, dst).WithTracking(1)
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			pubPush.receiveDelivered()
+		}()
+
+		wg.Add(1)
+		go func(status int, d chan telegraf.Metric) {
+			defer wg.Done()
+			if status == http.StatusNoContent {
+				for m := range d {
+					ro.AddMetric(m)
+					ro.Write()
+				}
+			}
+		}(test.status, dst)
+
+		http.HandlerFunc(pubPush.ServeHTTP).ServeHTTP(rr, req)
+
+		if status := rr.Code; status != test.status {
+			t.Errorf("%s - handler returned wrong status code: got %d want %d",
+				test.name, status, test.status)
+		}
+
+		if test.expected != "" {
```

Comment:
Not required but recommend using testify to avoid having to write so much, even if this is autogenerated with a snippet or such it will be more annoying to maintain vs `require.Equal(t, expected, actual)`